File: simple_server.py
#!/usr/bin/env python3
import http.server
import socketserver
import json
import time
import random
from urllib.parse import parse_qs, urlparse

class QuizHandler(http.server.SimpleHTTPRequestHandler):
    
    def log_message(self, format, *args):
        # Override to reduce verbose logging
        pass
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        self.end_headers()
    
    def do_GET(self):
        if self.path == '/api/health':
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            response = json.dumps({'status': 'ok'})
            self.wfile.write(response.encode())
            print("Health check OK")
            
        elif self.path.startswith('/api/questions/generate'):
            try:
                parsed_url = urlparse(self.path)
                params = parse_qs(parsed_url.query)
                
                question_type = params.get('type', ['math'])[0]
                difficulty = int(params.get('difficulty', ['1'])[0])
                
                if question_type == 'math':
                    num1 = random.randint(1, 20)
                    num2 = random.randint(1, 20)
                    op = random.choice(['+', '-'])
                    if op == '+':
                        answer = num1 + num2
                        q_text = f"{num1} + {num2} = ?"
                    else:
                        if num1 < num2:
                            num1, num2 = num2, num1
                        answer = num1 - num2
                        q_text = f"{num1} - {num2} = ?"
                    
                    question = {
                        'id': f"math_{int(time.time())}_{random.randint(1000, 9999)}",
                        'type': 'math',
                        'question': q_text,
                        'answer': str(answer),
                        'explanation': f"答えは {answer} です。"
                    }
                else:
                    # Language question
                    lang_questions = [
                        {'q': '「早い」の反対語は？', 'opts': ['遅い', '速い', '近い', '遠い', '高い'], 'ans': '遅い'},
                        {'q': '「大きい」の反対語は？', 'opts': ['小さい', '長い', '短い', '太い', '細い'], 'ans': '小さい'},
                        {'q': '「暑い」の反対語は？', 'opts': ['寒い', '冷たい', '涼しい', '暖かい', '熱い'], 'ans': '寒い'},
                        {'q': '「重い」の反対語は？', 'opts': ['軽い', '軟らかい', '硬い', '強い', '弱い'], 'ans': '軽い'},
                        {'q': '「新しい」の反対語は？', 'opts': ['古い', '若い', '新品', '綺麗', '汚い'], 'ans': '古い'}
                    ]
                    
                    selected = random.choice(lang_questions)
                    question = {
                        'id': f"lang_{int(time.time())}_{random.randint(1000, 9999)}",
                        'type': 'language',
                        'question': selected['q'],
                        'options': selected['opts'],
                        'answer': selected['ans'],
                        'explanation': f"正解は「{selected['ans']}」です。"
                    }
                
                # Store question
                if not hasattr(QuizHandler, 'questions'):
                    QuizHandler.questions = {}
                QuizHandler.questions[question['id']] = question
                
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps(question).encode())
                print(f"Generated {question_type} question: {question['question']}")
                
            except Exception as e:
                print(f"Error generating question: {e}")
                self.send_response(500)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': str(e)}).encode())
        
        elif self.path == '/' or self.path == '/index.html':
            try:
                with open('client/dist/index.html', 'r', encoding='utf-8') as f:
                    content = f.read()
                self.send_response(200)
                self.send_header('Content-type', 'text/html')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(content.encode())
            except:
                self.send_response(404)
                self.end_headers()
                self.wfile.write(b'Not found')
        else:
            super().do_GET()
    
    def do_POST(self):
        if self.path == '/api/users':
            try:
                content_length = int(self.headers['Content-Length'])
                data = json.loads(self.rfile.read(content_length))
                
                user = {
                    'id': f"user_{int(time.time())}_{random.randint(1000, 9999)}",
                    'name': data['name'],
                    'totalScore': 0,
                    'mathScore': 0,
                    'languageScore': 0,
                    'rank': '初心者',
                    'currentRank': 'bronze'
                }
                
                if not hasattr(QuizHandler, 'users'):
                    QuizHandler.users = {}
                QuizHandler.users[user['id']] = user
                
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps(user).encode())
                print(f"Created user: {user['name']}")
                
            except Exception as e:
                print(f"Error creating user: {e}")
                self.send_response(500)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': str(e)}).encode())
        
        elif self.path == '/api/answers':
            try:
                content_length = int(self.headers['Content-Length'])
                data = json.loads(self.rfile.read(content_length))
                
                question_id = data['questionId']
                user_answer = data['answer']
                user_id = data['userId']
                
                # Get question
                question = getattr(QuizHandler, 'questions', {}).get(question_id)
                if not question:
                    self.send_response(404)
                    self.send_header('Content-type', 'application/json')
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.end_headers()
                    self.wfile.write(json.dumps({'error': 'Question not found'}).encode())
                    return
                
                # Check answer
                is_correct = str(question['answer']).strip() == str(user_answer).strip()
                score_increment = 10 if is_correct else 0
                
                # Update user score
                user = getattr(QuizHandler, 'users', {}).get(user_id)
                if user:
                    user['totalScore'] += score_increment
                    if question['type'] == 'math':
                        user['mathScore'] += score_increment
                    else:
                        user['languageScore'] += score_increment
                    
                    # Update rank
                    if user['totalScore'] >= 100:
                        user['rank'] = 'エキスパート'
                        user['currentRank'] = 'gold'
                    elif user['totalScore'] >= 50:
                        user['rank'] = '中級者'
                        user['currentRank'] = 'silver'
                
                result = {
                    'id': f"answer_{int(time.time())}_{random.randint(1000, 9999)}",
                    'isCorrect': is_correct,
                    'scoreIncrement': score_increment
                }
                
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps(result).encode())
                print(f"Answer: {user_answer} -> {'Correct' if is_correct else 'Wrong'}")
                
            except Exception as e:
                print(f"Error processing answer: {e}")
                self.send_response(500)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': str(e)}).encode())
        
        else:
            super().do_POST()

if __name__ == "__main__":
    PORT = 3001
    print(f"Starting server on port {PORT}...")
    with socketserver.TCPServer(('0.0.0.0', PORT), QuizHandler) as httpd:
        print(f"Server running at http://localhost:{PORT}")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nServer stopped")

File: simple-server.js
const express = require('express');
const path = require('path');
const fs = require('fs');

const app = express();
const port = 3000;

// Middleware
app.use(express.json());
app.use(express.static(path.join(__dirname, 'client/dist')));

// Simple in-memory storage
let users = [];
let questions = [];
let answers = [];

// API Routes
app.get('/api/health', (req, res) => {
  res.json({ status: 'OK' });
});

app.post('/api/users', (req, res) => {
  const user = {
    id: `user_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
    ...req.body,
    createdAt: new Date()
  };
  users.push(user);
  res.json(user);
});

app.get('/api/users/:id', (req, res) => {
  const user = users.find(u => u.id === req.params.id);
  if (!user) {
    return res.status(404).json({ error: 'User not found' });
  }
  res.json(user);
});

app.put('/api/users/:id', (req, res) => {
  const userId = req.params.id;
  const userIndex = users.findIndex(u => u.id === userId);
  
  if (userIndex === -1) {
    return res.status(404).json({ error: 'User not found' });
  }
  
  users[userIndex] = { ...users[userIndex], ...req.body };
  res.json(users[userIndex]);
});

app.post('/api/questions', (req, res) => {
  const question = {
    ...req.body,
    createdAt: new Date()
  };
  questions.push(question);
  res.json(question);
});

app.post('/api/answers', (req, res) => {
  const { userId, questionId, userAnswer } = req.body;
  
  console.log('Answer submission request:', { userId, questionId, userAnswer });
  console.log('Total users:', users.length);
  console.log('Total questions:', questions.length);
  
  // Find user and question
  const user = users.find(u => u.id === userId);
  const question = questions.find(q => q.id === questionId);
  
  console.log('User found:', !!user);
  console.log('Question found:', !!question);
  
  if (!user) {
    console.log('User not found with ID:', userId);
    return res.status(404).json({ error: 'User not found' });
  }
  
  if (!question) {
    console.log('Question not found with ID:', questionId);
    console.log('Available questions:', questions.map(q => q.id));
    return res.status(404).json({ error: 'Question not found' });
  }
  
  // Check if answer is correct
  const isCorrect = userAnswer.toString().trim() === question.answer.toString().trim();
  
  // Create answer record
  const answer = {
    id: `answer_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
    userId,
    questionId,
    userAnswer,
    isCorrect,
    answeredAt: new Date()
  };
  answers.push(answer);
  
  // Update user scores
  if (isCorrect) {
    user.totalScore = (user.totalScore || 0) + 1;
    if (question.type === 'math') {
      user.mathScore = (user.mathScore || 0) + 1;
    } else if (question.type === 'language') {
      user.languageScore = (user.languageScore || 0) + 1;
    }
    
    // Update user in the users array
    const userIndex = users.findIndex(u => u.id === userId);
    if (userIndex !== -1) {
      users[userIndex] = user;
    }
  }
  
  // Mock rewards
  const newRewards = [];
  if (user.totalScore === 1) {
    newRewards.push({
      id: 'reward_1',
      name: 'はじめの一歩',
      description: '最初の正解！',
      requiredScore: 1,
      icon: '🎉'
    });
  }
  
  res.json({
    answer,
    user,
    scoreIncrement: isCorrect ? 1 : 0,
    rankUp: false,
    newRewards
  });
});

// Serve frontend
app.get('/', (req, res) => {
  res.sendFile(path.join(__dirname, 'client/dist/index.html'));
});

app.listen(port, () => {
  console.log(`Server running on port ${port}`);
  console.log(`Web app: http://localhost:${port}`);
  console.log(`API: http://localhost:${port}/api`);
});